CleverRaven/Cataclysm-DDA
# Weapon Proficiency Support

#### Summary
Features "Weapon Proficiency"

#### Purpose of change
As it stands, weapon selection in Catacylsm is a choice between the three core damage types of Bash, Slash, and Stab, with the player either selecting whatever does the most damage in the area the have the highest skill, or to choose a sub-optimal weapon for roleplay purposes. This leads to a lack of nuance in weapon selection that isn't as interesting to engage with as it could be.

#### Describe the solution
This adds a system that adds weapon proficiencies for each melee weapon type across three tier of proficiency. A player can now gain category familiarity, category proficiency, and category mastery for each category.  It adds EoCs to allow an increase in each weapon category proficiency, and does so by:

1. Proc the EoCs on a successful hit on either a Monster or a Character
2. Compare each weapon category to the wielded weapon
2.1 On a match with the wielded weapon, check the highest tier of proficiency the player has  in that category
2.2 After finding the highest tier, increment the amount of experience in the relevant tier of proficiency unless already a master
2.3 Compare the lowest tier of proficiency determined so far to the current tier. If the current tier is lower, set lowest tier to current tier.
3. After each category is checked, refund a small amount of stamina spent on the swinging the weapon based on the players lowest matching category proficiency
3.a Familiarity provides a 2.5% refund
3.b Proficiency provides a 5% refund
3.c Mastery provides a 10% refund

#### Describe alternatives you've considered
Other alternatives to this have been tracking proficiency only in the specific item, providing a to-hit bonus based on proficiency, an attack move cost discount, or just leaving the proficiency frame work as it is with no benefit to provide support to later additions.

#### Testing
1. New Character
2. Spawn dummies
3. Spawn weapon of appropriate category
4. Attack dummies
5. Check proficiency gain
6. Debug set proficiency to each tier
7. Attack dummies
8. Observe if results match expectation

#### Additional context
While functional, the descriptions for both the practice recipes and the proficiencies themselves are kinda dull, suggestions are well appreciated.
